backends/opt/ceed-opt-tensor.c

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2017-2018, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory. LLNL-CODE-734707.
+// All Rights reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed/ceed.h>
+#include <ceed/backend.h>
+#include "ceed-opt.h"
+
+//------------------------------------------------------------------------------
+// Tensor Contract Blocked
+//------------------------------------------------------------------------------
+int CeedTensorContractApply_Blocked_Opt(CeedTensorContract contract, CeedInt A,
+                                        CeedInt B, CeedInt C, CeedInt J,
+                                        const CeedScalar *restrict t,
+                                        CeedTransposeMode t_mode, const CeedInt add,
+                                        const CeedScalar *restrict u,
+                                        CeedScalar *restrict v, const CeedInt CC) {
+  CeedInt t_stride_0 = B, t_stride_1 = 1;
+  if (t_mode == CEED_TRANSPOSE) {
+    t_stride_0 = 1; t_stride_1 = J;
+  }
+
+  for (CeedInt a=0; a<A; a++)
+    for (CeedInt b=0; b<B; b++)
+      for (CeedInt j=0; j<J; j++) {
+        CeedScalar tq = t[j*t_stride_0 + b*t_stride_1];
+        for (CeedInt c=0; c<(C/CC)*CC; c+=CC) // unroll
+          for (CeedInt cc=0; cc<CC; cc++)
+            v[(a*J+j)*C+(c+cc)] += tq * u[(a*B+b)*C+(c+cc)];
+        CeedInt c = (C/CC)*CC;
+        for (CeedInt cc=0; cc<C-c; c++)
+          v[(a*J+j)*C+(c+cc)] += tq * u[(a*B+b)*C+(c+cc)];
+      }
+
+  return CEED_ERROR_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Tensor Contract Serial
+//------------------------------------------------------------------------------
+int CeedTensorContractApply_Serial_Opt(CeedTensorContract contract, CeedInt A,
```

Comment:
While here this does better than the ref tensor contract `DoFs/Sec in CG: 5.69806 (5.69806) million` vs `DoFs/Sec in CG: 1.30158 (1.30158) million`